include/revng/Pipeline/Containers.h

```diff
@@ -0,0 +1,227 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstddef>
+#include <fstream>
+#include <memory>
+#include <optional>
+#include <set>
+#include <sstream>
+#include <string>
+
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+
+#include "revng/Pipeline/ContainerFactory.h"
+#include "revng/Pipeline/InputOutputContract.h"
+#include "revng/Pipeline/Target.h"
+#include "revng/Support/Assert.h"
+
+namespace Pipeline {
+
+/// A wrapper around a optional<container> and the pointer to the factory that
+/// can spawn this container when needed
+class MaybeContainer {
```

Comment:
* [ ] If it's self instantiating it's more like `class LazyContainer`.
* [ ] Are you sure you want this? This approach leads you to duplicate the list of methods. Usually you have a registry and you have a `getOrCreate` method instead of `MaybeContainer`. The only advantage I can see in the approach we currently have here is if you want to be able to pass this around, but please consider if this is pushing lazyness a lot for little/no return.